catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergApiV1GenericResource.java

```diff
@@ -97,31 +110,95 @@ public IcebergConfigResponse getConfig(
+    DecodedPrefix decodedPrefix = decodePrefix(prefix);
+    ContentKey key = ContentKey.fromPathString(identifier);
+    String requestedS3Uri = S3Utils.asS3Location(request.uri());
+
+    return collectBasePrefixes(decodedPrefix, key)
+        .filter(requestedS3Uri::startsWith)
+        .map(ignored -> doRequestSign(request, decodedPrefix, key))
+        .toUni()
+        .replaceIfNullWith(() -> signingRequestDenied(request));
+  }
+
+  private Multi<String> collectBasePrefixes(DecodedPrefix decodedPrefix, ContentKey key)
+      throws NessieNotFoundException {
+    try {
+      ParsedReference ref = decodedPrefix.parsedReference();
+      ContentResponse contentResponse =
+          nessieApi
+              .getContent()
+              .refName(ref.name())
+              .hashOnRef(ref.hashWithRelativeSpec())
+              .getSingle(key);
+
+      // table exists: fetch snapshot, get additional locations
+      CompletionStage<SnapshotResponse> stage =
+          catalogService.retrieveSnapshot(
+              SnapshotReqParams.forSnapshotHttpReq(ref, "iceberg", null),
+              key,
+              new CatalogUriResolverImpl(uriInfo),
+              contentResponse.getContent().getType());
+      return Uni.createFrom()
+          .completionStage(stage)
+          .map(SnapshotResponse::nessieSnapshot)
+          .onItem()
+          .<String>transformToMulti(
+              snapshot ->
+                  Multi.createFrom()
+                      .emitter(
+                          e -> {
+                            e.emit(snapshot.icebergLocation());
+                            snapshot.additionalKnownLocations().forEach(e::emit);
+                            e.complete();
+                          }))
+          // allow the parent directory of each iceberg location,
+          // assuming the locations are of the form: <base-prefix>/<random-part>
+          .map(location -> location.substring(0, location.lastIndexOf('/') + 1));
+
+    } catch (NessieContentNotFoundException e) {
+      // table does not exist: allow any path under the base prefix
+      WarehouseConfig warehouse = catalogConfig.getWarehouse(decodedPrefix.warehouse());
+      String basePrefix = icebergBasePrefix(warehouse.location(), key);
+      return Multi.createFrom().item(basePrefix);
+    }
+  }
 
+  private Response doRequestSign(
+      IcebergS3SignRequest request, DecodedPrefix decodedPrefix, ContentKey key) {
+    URI uri = URI.create(request.uri());
     Optional<String> bucket = s3options.extractBucket(uri);
     Optional<String> body = Optional.ofNullable(request.body());
 
     SigningRequest signingRequest =
         SigningRequest.signingRequest(
             uri, request.method(), request.region(), bucket, body, request.headers());
 
-    SigningResponse signed = signer.sign(ref.name(), identifier, signingRequest);
+    SigningResponse signed =
+        signer.sign(decodedPrefix.parsedReference().name(), key.toPathString(), signingRequest);
```

Comment:
I noticed that the current impl of this component ignores the reference and the key, FYI.

Reply:
Yes, the original idea at that time (not fully thought through) was that the signer implementation checks reference and key. But that's no longer the case, so I guess the parameters can go away.